### PR TITLE
Adaptive-step Runge Kutta Example

### PR DESCRIPTION
**Pull request by duvenaud**
I ported over the [jax implementation of Dopri5](https://github.com/google/jax/blob/master/jax/experimental/ode.py) to Dex.  Its error gets arbitrarily small as you tighten the error tolerances, but for some reason its output doesn't match that of the Jax demo.  More on that below.  Notes on writing this:

1. Having the type system catch lots of errors was nice, because it let me mostly debug one function at a time as I wrote it.  The type system caught two redundant ops that I had mistakenly included when I wrote the original jax version.  They were left over from the hand-vmapped pytorch version by Ricky Chen that I based it on.
2. The type system reduced the need for comments.  About half the comments in jax's ```odeint``` simply list the types of arguments.  However  I wasn't sure how often to annotate inner functions with types.  Whenever I had a top-level type mismatch, I annotated the inner functions' types to narrow down the error.  However, for brevity and flexibility I removed them after and it might have hurt readability.  Notably, I never had to write the type of the complex solver state tuple.
3. The generic ```zero``` function is really nice to avoid writing sizes when initializing.  I would have also liked an analogous ```nans``` function, but I couldn't figure out how to write an equally powerful argumentless ```nans``` function without polluting the typeclasses.
4. Dex's support for triangular arrays means it could skip some redundant ops in the inner loop inherited from the Jax and PyTorch versions where they multiplying with zero-padded arrays.

Unfortunately there is some sort of mismatch with the Jax version, and this version ends up having slightly higher error on my simple test function.  I compared everything line by line twice and can't find any algorithmic difference.  The RK steps in Jax and Dex give slightly different answers in the last printed decimal place.  I think this might just be a float32 v float64 issue but I don't really know how to control float32 v 64 in both Jax and Dex, so that's where I left it for now.

Dex RK step:
```
:p runge_kutta_step myDyn z0 f0 t0 1.0
( [2.7183332]
, ( [2.7183332]
, ( [-3.5e-4]
, [[1.0], [1.2], [1.345], [2.28], [2.5863924], [2.8436363], [2.7183332]] ) ) )
```

Jax RK step:

```
(DeviceArray([2.7183337], dtype=float32), DeviceArray([2.7183337], dtype=float32), DeviceArray([-0.00035], dtype=float32), DeviceArray([[1.       ],
             [1.2      ],
             [1.345    ],
             [2.2800002],
             [2.5863914],
             [2.8436356],
             [2.7183337]], dtype=float32))
```

### CODE REVIEW

**Comment by apaszke** [comment]
Awesome! I'm curious how does the perf compare between Dex and JAX. Did you check that?

**Comment by duvenaud** [comment]
Noe, I haven't done any comparisons yet.  Although I think the first commit does the exact same ops as the Jax version per-step, so it'd be a good comparison.

**Comment by duvenaud** [comment]
Okay, thanks to @apaszke 's help I managed to get a lower-triangular tableaux working without any zero padding.  Specifically, the main inner loop now looks like the textbook definition:

```
  runge_kutta_stage = \i func_evals.
    ti = t0 + dt .* alpha.i
    zi = z0 + dt .* dot beta.i func_evals
    func zi ti
```

I still don't know why the error achieved isn't quite as good as the Jax version, but at least it's taking advantage of Dex's extra flexibility now.  I'm not planning to work on this PR any more until AD is stable again and we can think about incorporating the reverse-time adjoint sensitivity method!

File: examples/ode-integrator.dx [hunk 1]
```diff
@@ -1,60 +1,180 @@
...
+  --  Returns:
+  --    Values of the solution at each time point in times.
+  rtol = 1.4e-8 -- relative local error tolerance for solver.
+  atol = 1.4e-8 -- absolute local error tolerance for solver.
+  max_iters = 10000
+
+  integrate_to_next_time = \iter init_carry.
+    target_t = times.iter
+
+    stopping_condition = \(_, _, t, dt, _, _).
```
Comment (apaszke) [inline examples/ode-integrator.dx:121]: How about using extensible records for the solver state? That would make unpacking like this much more readable I think

Comment (duvenaud) [reply examples/ode-integrator.dx:122]: I tried making this change, and it does make this part more readable, but it made the rest of the function much more verbose.  However I'm not sure I have the hang of extensible records yet.  Perhaps we could merge it and then @hexahedria could update this example as a demo of extensible records?

File: examples/ode-integrator.dx [hunk 2]
```diff
@@ -1,60 +1,179 @@
...
+This version is a port of the [Jax implementation](https://github.com/google/jax/blob/4236eb2b5929b9643977553f7f988ca518b7df4e/jax/experimental/ode.py).
+A notable algorithmic difference is that it uses a lower-triangular
+matrix type for the Butcher tableua, and so avoid zero-padding everywhere.
+
+Time = Float
+
+-- Should these go in the prelude?
+def length (x: d=>Float) : Float = sqrt $ sum for i. sq x.i
+def (./) (x: d=>Float) (y: d=>Float) : d=>Float = for i. x.i / y.i
+def dot (_:VSpace v) ?=> (s:d=>Float) (vs:d=>v) : v = sum for j. s.j .* vs.j
```
Comment (apaszke) [inline examples/ode-integrator.dx:12]: This one is already in the prelude as `vdot`. I think we could move `length` there, although maybe it would be clearer to call it `norm` or `norm2` just so that it's harder to confuse with `size`.

File: examples/ode-integrator.dx [hunk 3]
```diff
@@ -1,60 +1,179 @@
...
+
+def runge_kutta_step (_:VSpace v) ?=> (func:v->Time->v)
+    (z0:v) (f0:v) (t0:Time) (dt:Time) : (v & v & v & (Fin 7)=>v) =
+
+  runge_kutta_stage = \i func_evals.
+    ti = t0 + dt .* alpha.i
+    zi = z0 + dt .* dot beta.i func_evals
+    func zi ti
+
+  evals_init = for i. select (i == (0@_)) f0 zero
```
Comment (apaszke) [inline examples/ode-integrator.dx:74]: Wouldn't it be simpler to initialize the state with `zero`s and then set the first index to `f0`?

File: examples/ode-integrator.dx [hunk 4]
```diff
@@ -1,60 +1,179 @@
...
+
+  evals_filled = snd $ withState evals_init \func_evals. for i:(Fin 6).
+    cur_evals = for j:(..i). (get func_evals).((ordinal j)@_)
```
Comment (apaszke) [inline examples/ode-integrator.dx:77]: nit: we don't do too many optimizations around effects, and I wouldn't be surprised if `get ref!idx` would be _a lot_ faster than `(get ref).idx` (with the second creating a copy of the state, only to index out a single value).

File: examples/ode-integrator.dx [hunk 5]
```diff
@@ -1,60 +1,179 @@
...
+    ft = runge_kutta_stage i cur_evals
+    func_evals!(((ordinal i) + 1)@_) := ft
```
Comment (apaszke) [inline examples/ode-integrator.dx:79]: This is a great example for some dynamic programming utilities and triangular types. We should think about exposing something better. We could have a combinator such as `dynamic :: (n : Type) ?-> (i:n -> (..i)=>a -> a) -> n=>a`, and we could compile it specially such that it operates on an array in-place, and keeps appending to it after every iteration.

File: examples/ode-integrator.dx [hunk 6]
```diff
@@ -1,60 +1,179 @@
...
+  z_last =  z0 + dt .* dot c_sol   evals_filled
+  z_last_error = dt .* dot c_error evals_filled
+  f_last = evals_filled.(6@_)
+  (z_last, f_last, z_last_error, evals_filled)
+
+def error_ratio (error_estimates:d=>Float) (rtol:Float) (atol:Float)
+                (z0:d=>Float) (z1:d=>Float) : Float =
+  err_tols = for i. atol + rtol * (max (abs z0.i) (abs z1.i))
+  err_ratios = error_estimates ./ err_tols
+  mean for i. sq err_ratios.i
```
Comment (apaszke) [inline examples/ode-integrator.dx:90]: nit: even though this might parse correctly, I usually write `mean $ for i. ...` just to clarify where the parens are

File: examples/ode-integrator.dx [hunk 7]
```diff
@@ -1,60 +1,179 @@
...
+           125. / 192. - 451. / 720., -2187. / 6784. + 12231. / 42400.,
+           11. / 84. - 649. / 6300., -1. / 60.]
+
+def runge_kutta_step (_:VSpace v) ?=> (func:v->Time->v)
+    (z0:v) (f0:v) (t0:Time) (dt:Time) : (v & v & v & (Fin 7)=>v) =
+
+  runge_kutta_stage = \i func_evals.
+    ti = t0 + dt .* alpha.i
+    zi = z0 + dt .* dot beta.i func_evals
+    func zi ti
```
Comment (apaszke) [inline examples/ode-integrator.dx:72]: Is there any reason to avoid inlining this function?

File: examples/ode-integrator.dx [hunk 8]
```diff
@@ -1,60 +1,179 @@
...
 
-:p odeint myDyn z0 t0 t1 dt
-> [[2.7169456]]
+approx_e = odeint myDyn z0 t0 t1
+:p approx_e
+> [[2.7201762]]
+
+exact_e = [[exp 1.0]]
+
+:p (approx_e - exact_e) -- amount of numerical error
```
Comment (apaszke) [inline examples/ode-integrator.dx:170]: nit: we could use `~~` now!

**Review by apaszke — APPROVED** [review]

File: examples/ode-integrator.dx (hunk 2 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:12]: The one in the prelude is
```
def vdot (x:n=>Float) (y:n=>Float) : Float = fsum \i. x.i * y.i
```
which I didn't want to convert to the generic vector version since then it couldn't use the faster ```fsum```.

Since ```fsum``` is slated for unification with ```sum```, how about I put my ```dot``` in the prelude, and we can delete ```vdot``` later when ```fsum``` is made obsolete?

File: examples/ode-integrator.dx (hunk 7 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:72]: No, but since this function has a name in the ODE literature it was nice to name it explicitly.   It's also left over from copying the Jax implementation, where the equivalent function looks a little hairier:

```
  def body_fun(i, k):
    ti = t0 + dt * alpha[i-1]
    yi = y0 + dt * jnp.dot(beta[i-1, :], k)
    ft = func(yi, ti)
    return k.at[i, :].set(ft)
```

File: examples/ode-integrator.dx (hunk 3 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:74]: Here's the simplest way I know how to write that:
```
evals_init = snd $ withState zero \r.
  r!(0@_) := f0
```
I'll put that version in for now, although it's not all that readable either imo.

File: examples/ode-integrator.dx (hunk 7 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:72]: Anyways inlining saved 3 lines, so I inlined it, thanks.

File: examples/ode-integrator.dx (hunk 4 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:77]: Ah, thanks for the tip, I don't think I appreciated the difference back when I wrote this.  Fixed.

As for ```sliceUpTo```, I'm not sure how I could use it here, since I need to index into a reference.

File: examples/ode-integrator.dx (hunk 5 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:79]: Yeah, it felt a bit disappointing to initialize an array with fake values and then fill it, like a caveman.  Ideally we'd never use a placeholder value anywhere (before compile time).

File: examples/ode-integrator.dx (hunk 6 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:90]: Point taken.

File: examples/ode-integrator.dx (hunk 8 above)
Comment (duvenaud) [reply examples/ode-integrator.dx:170]: I would, except I wanted to show the amount of error.  Actually, what I should really show is a graph of error vs tolerance, to show the convergence rate.

File: examples/ode-integrator.dx (hunk 2 above)
Comment (apaszke) [reply examples/ode-integrator.dx:12]: Oh, we should remove `fsum`. I've already implemented the inlining optimization necessary to make `sum` just as fast!

File: examples/ode-integrator.dx (hunk 3 above)
Comment (apaszke) [reply examples/ode-integrator.dx]: Oh, I meant fusing the assignment into the subsequent `withState` block. So instead of doing
```
evals_init = for i. select (i == (0@_)) f0 zero
withState evals_init \evals.
  ...
```
you would do
```
withState zero \evals.
  evals!(0@_) := f0
  ...
```
which I think is nicer?

File: examples/ode-integrator.dx (hunk 4 above)
Comment (apaszke) [reply examples/ode-integrator.dx]: Yeah good point... It's so annoying that we have to duplicate all the functions to work on refs...

Comment (apaszke) [reply examples/ode-integrator.dx]: Though if we added a type class called "indexable" which allows you to do `.` on a type, then I think we could implement a generic function to do that!

File: examples/ode-integrator.dx (hunk 8 above)
Comment (apaszke) [reply examples/ode-integrator.dx:167]: Ohh that would cool!

**Review by apaszke — APPROVED** [review]

File: examples/ode-integrator.dx (hunk 3 above)
Comment (duvenaud) [reply examples/ode-integrator.dx]: Oh, of course!  Yes, I'll fix that next time I touch that code.

It's funny, because I learned how to use ```withState``` just by looking at examples, it didn't occur to me until this conversation that you could decouple it from a ```for``` loop.  I was thinking of it as just something you add to a for loop to let it make mutating updates.

Comment (apaszke) [reply examples/ode-integrator.dx]: Right, you can definitely use `withState` without loops, although it usually doesn't make sense, because if you have the code unrolled then you don't need the state effect! But it's good to be able to e.g. do this initialization outside of the loop.
